CRM/PivotReport/AbstractData.php

```diff
@@ -0,0 +1,526 @@
+      }
+      return $valueArray;
+    }
+
+    if (!empty($this->formattedValues[$key][$value])) {
+      return $this->formattedValues[$key][$value];
+    }
+
+    if (!empty($this->fields[$key]['customField'])) {
+      switch ($this->fields[$key]['customField']['data_type']) {
+        case 'File':
+          $result = CRM_Utils_System::formatWikiURL($value['fileURL'] . ' ' . $value['fileName']);
+          $this->formattedValues[$key][$value] = $result;
+          return $result;
+        break;
+        // For few field types we can use 'formatCustomValues()' core method.
+        case 'Date':
+        case 'Boolean':
+        case 'Link':
+        case 'StateProvince':
+        case 'Country':
+          $data = array('data' => $value);
+          CRM_Utils_System::url();
+          $result = CRM_Core_BAO_CustomGroup::formatCustomValues($data, $this->fields[$key]['customField']);
+          $this->formattedValues[$key][$value] = $result;
+          return $result;
+        break;
+        // Anyway, 'formatCustomValues()' core method doesn't handle some types
+        // such as 'CheckBox' (looks like they aren't implemented there) so
+        // we deal with them automatically by custom handling of 'optionValues' array.
+      }
+    }
+
+    if (!empty($this->fields[$key]['optionValues'])) {
+      $result = $this->fields[$key]['optionValues'][$value];
+      $this->formattedValues[$key][$value] = $result;
+      return $result;
+    }
+
+    $result = strip_tags($this->customizeValue($key, $value));
+    $this->formattedValues[$key][$value] = $result;
+
+    return $result;
+  }
+
+  /**
+   * Additional function for customizing Entity value by its key
+   * (if it's needed). For example: we want to return Campaign's title
+   * instead of ID.
+   *
+   * @param string $key
+   *   Field key
+   * @param string $value
+   *   Field value
+   *
+   * @return string
+   */
+  protected function customizeValue($key, $value) {
+    if (!empty($this->customizedValues[$key][$value])) {
+      return $this->customizedValues[$key][$value];
+    }
+
+    $result = $value;
+
+    $this->customizedValues[$key][$value] = $result;
+
+    return $result;
+  }
+
+  /**
+   * Returns an empty row containing field names as keys.
+   *
+   * @return array
+   */
+  protected function getEmptyRow() {
+    $result = array();
+
+    foreach ($this->fields as $key => $value) {
+      if (!empty($value['title'])) {
+        $key = $value['title'];
+      }
+      $result[$key] = '';
+    }
+
+    return $result;
+  }
+
+  /**
+   * Returns an array containing all Fields and Custom Fields of entity,
+   * keyed by their API keys and extended with available fields Option Values.
+   *
+   * @return array
+   */
+  abstract protected function getFields();
+
+  /**
+   * Gets total number of entities.
+   *
+   * @param array $params
+   *
+   * @return int
+   */
+  abstract protected function getCount(array $params);
```

Comment:
I see that many of the methods from this class have been overwritten by DataActivity (`customizeValues`, `formatResult` or `splitMultiValues` for example), which makes me wonder if these methods shouldn't have  been marked as abstract instead or if there's something that is duplicated between the two classes

Reply:
I left these methods implementation as they may be useful for other (non-Activity) entities as the  implementations should work fine with plain entities data.

Reply:
@coldrunKacper that's definitely a good reason for keeping them, but I think that they need to be improved a little bit to actually become reusable by other child classes. Without doing that, I fear that almost all the child classes will end up duplicating a lot of code just like DataActivity.

Here are some suggestions of what can be improved on these methods:
`splitMultiValue()`
This is a comparison between the classes: https://www.diffchecker.com/r4e9gjrT
Basically, the only difference here is how we calculate the index. This can be solved by adding a `getEntityIndex()` method in AbstractData and then overriding it in child classes that need a different logic to get the index

`formatResult()`
This is a comparison between the classes: https://www.diffchecker.com/2pHUtman
For this one, perhaps adding a `postFormatResult()` method which child classes can override in order to add/change things on the result

`customizeValue()`
For this one, I think we simply drop the implementation from `AbstractData` and make the method there abstract. Maybe I'm wrong, but I implementation of this method simply doesn't make sense to me. You're simply storing `$value` in an array and then returning it. What is the purpose of that?

`rebuildCache()`
This is a comparison between the classes: https://www.diffchecker.com/dkhfLSFJ
The duplication can be removed if you override `getEmptyRow()` in DataActivity to return all the necessary rows, including the three custom ones

Reply:
I've implemented the changes. Regarding formatResult() and rebuildCache() methods I've created a new class's array $additionalHeaderFields to store additional custom fields instead of defining it in these two methods.